### PR TITLE
gh-125498: Update JIT builds to use LLVM 19 and use `preserve_none`

### PR DESCRIPTION
**Pull request by savannahostrowski**
* Issue: gh-125498
* Issue: gh-123714

### CODE REVIEW

File: Tools/jit/jit_definitions.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#ifndef JIT_DEFINITIONS_H
```
Comment (savannahostrowski) [inline Tools/jit/jit_definitions.h]: I couldn't really find a naming convention precedent here since this is the first header file in `Tools/jit.`  I opted not to include `preserve_none` in the name here since maybe we'll want to include more in the file in the future.

Happy to rename :)

**Comment by savannahostrowski** [comment]
CI failures seem to be related to https://github.com/python/cpython/issues/125521

File: Tools/jit/jit_definitions.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+#ifndef Py_BUILD_CORE
+#  error "this header requires Py_BUILD_CORE define"
+#endif
+
+#ifdef _Py_JIT
+
+// To use preserve_none in JIT builds, we need to declare a separate function pointer with __attribute__((preserve_none)) since this attribute is not supported in < clang 19.
+// This is functionally the same as jit_func_native from Include/internal/pycore_jit.h except that it has the __attribute__((preserve_none)) attribute.
+typedef _Py_CODEUNIT *(*jit_func_preserve_none)(_PyInterpreterFrame *frame, _PyStackRef *stack_pointer, PyThreadState *tstate) __attribute__((preserve_none));
```
Comment (brandtbucher) [inline Tools/jit/jit_definitions.h]: This header only needs these three lines. We don't need all of the copy-pasted stuff at the top and bottom of this file, since it's not part of the normal interpreter headers (the context in which we're using this is a lot narrower).

Also, we probably want to pass `-Wno-ignored-attributes` to Clang on `i686-pc-windows-msvc`, since we get warnings that `preserve_none` isn't supported there.

File: Tools/jit/_targets.py [hunk 3]
```diff
@@ -145,41 +141,7 @@ async def _compile(
             "-std=c11",
             *self.args,
         ]
-        if self.ghccc:
```
Comment (brandtbucher) [inline Tools/jit/_targets.py:148]: Good riddance!

File: Tools/jit/jit_definitions.h (hunk 1 above)
Comment (brandtbucher) [reply Tools/jit/jit_definitions.h]: How about just `Tools/jit/jit.h`?

**Review by brandtbucher — COMMENTED** [review]
Awesome! Should I run benchmarks?

Initial review of these changes below. Also, this should probably have its own blurb.

File: Tools/jit/trampoline.c [hunk 4]
```diff
@@ -4,9 +4,8 @@
 #include "pycore_frame.h"
 #include "pycore_jit.h"
 
-// This is where the calling convention changes, on platforms that require it.
-// The actual change is patched in while the JIT compiler is being built, in
-// Tools/jit/_targets.py. On other platforms, this function compiles to nothing.
+#include "jit_definitions.h"
+
 _Py_CODEUNIT *
 _ENTRY(_PyInterpreterFrame *frame, _PyStackRef *stack_pointer, PyThreadState *tstate)
```
Comment (brandtbucher) [inline Tools/jit/trampoline.c]: The only reason this name was different was because of the GHC hack. Now we can get rid of `entry_symbol` (and the associated comment) in `_targets.py` and just use the string `"_JIT_ENTRY"` in the assert on the following line.

```suggestion
_JIT_ENTRY(_PyInterpreterFrame *frame, _PyStackRef *stack_pointer, PyThreadState *tstate)
```

File: Tools/jit/jit_definitions.h (hunk 2 above)
Comment (savannahostrowski) [reply Tools/jit/jit_definitions.h]: Would it be worth keeping the include guard, since we have this header referenced in both trampoline.c and template.c?

Comment (brandtbucher) [reply Tools/jit/jit_definitions.h]: That's to guard against the same header being included multiple times in one C file (so, the oppositite situation). I don't think we need to worry about that.

**Comment by savannahostrowski** [comment]
Once CI passes, yes! Let's run the benchmarks. Thanks for the review!

**Comment by mdboom** [comment]
Can you merge in main to get [this fix in](https://github.com/python/cpython/pull/125477)?  That way we can benchmark on Windows as well.

**Comment by savannahostrowski** [comment]
@mdboom Done!

**Comment by savannahostrowski** [comment]
So it looks like the Python version installed with `brew` is failing transitively...doing a bit of spelunking here.

**Comment by savannahostrowski** [comment]
Okay, after a bit of digging here, my cursory diagnosis around the failing x86 macOS runs is essentially this (and I am by no means a Homebrew or Ruby expert):
- The default Python version of Homebrew was recently updated to 3.13 (see https://github.com/Homebrew/homebrew-core/pull/182840). As part of this PR, the formulae for Python 3.13 seem to vastly differ from 3.12...and 3.12 was also updated.
It seems like Homebrew is no longer handling overwriting files in the way it once did. More specifically, the combo of the `setup-python` step and then `brew` installing Python (via LLVM) is not cooperating. I did some testing here, and it looks like this issue only happens with 3.12 (preliminary run of https://github.com/savannahostrowski/test_repo_python_install/actions/runs/11453722971 but waiting on the `fail fast` run for the full output).

I'll likely file an issue on Homebrew in the morning.

File: Tools/jit/_targets.py [hunk 5]
```diff
@@ -145,41 +137,7 @@ async def _compile(
             "-std=c11",
             *self.args,
         ]
```
Comment (brandtbucher) [inline Tools/jit/_targets.py:142]: We can combine these. Then you can delete the following line, and just use `args` instead of `args_o` in the `_llvm.run` call.

```suggestion
            "-o",
            f"{o}",
            f"{c}",
            *self.args,
        ]
```

File: Include/internal/pycore_jit.h [hunk 6]
```diff
@@ -11,7 +11,7 @@ extern "C" {
 
 #ifdef _Py_JIT
 
-typedef _Py_CODEUNIT *(*jit_func)(_PyInterpreterFrame *frame, _PyStackRef *stack_pointer, PyThreadState *tstate);
```
Comment (brandtbucher) [inline Include/internal/pycore_jit.h]: Let's leave this as `jit_func` (and the new one as `jit_func_preserve_none`). I think the native calling convention can be implied.

File: Tools/jit/jit.h [hunk 7]
```diff
@@ -0,0 +1,3 @@
+// To use preserve_none in JIT builds, we need to declare a separate function pointer with __attribute__((preserve_none)) since this attribute is not supported in < clang 19.
+// This is functionally the same as jit_func_native from Include/internal/pycore_jit.h except that it has __attribute__((preserve_none)).
+typedef _Py_CODEUNIT *(*jit_func_preserve_none)(_PyInterpreterFrame *frame, _PyStackRef *stack_pointer, PyThreadState *tstate) __attribute__((preserve_none));
```
Comment (brandtbucher) [inline Tools/jit/jit.h]: I think this works, and it means we don't need to keep the two signatures in sync! It basically says "`jit_func_preserve_none` is `jit_func`, but with `__attribute__((preserve_none))`":

```suggestion
typedef jit_func __attribute__((preserve_none)) jit_func_preserve_none;
```

File: Tools/jit/jit.h [hunk 8]
```diff
@@ -0,0 +1,3 @@
+// To use preserve_none in JIT builds, we need to declare a separate function pointer with __attribute__((preserve_none)) since this attribute is not supported in < clang 19.
+// This is functionally the same as jit_func_native from Include/internal/pycore_jit.h except that it has __attribute__((preserve_none)).
```
Comment (brandtbucher) [inline Tools/jit/jit.h]:
```suggestion
// To use preserve_none in JIT builds, we need to declare a separate function
// pointer with __attribute__((preserve_none)), since this attribute may not be
// supported by the compiler used to build the rest of the interpreter.
```

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-22-04-18-53.gh-issue-125498.cFjPIn.rst [hunk 9]
```diff
@@ -0,0 +1,6 @@
+The JIT has been updated to leverage LLVM 19’s preserve_none attribute, which
+adds native support for the Glasgow Haskell Compiler Calling Convention (ghccc)
+for both x86-64 and AArch64 targets. This removes the need to manually patch
+the calling convention in LLVM IR, simplifying the JIT compilation process.
+Additionally, this update enables the optimization across all platforms, except
+for i686-pc-windows-msvc where the attribute is not supported.
```
Comment (brandtbucher) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-10-22-04-18-53.gh-issue-125498.cFjPIn.rst]: Hm, I'd reword this. The new calling convention is actually different from the GHC calling convention. And I don't think we need to mention 32-bit Windows.

How about this?

```suggestion
The JIT has been updated to leverage Clang 19’s new ``preserve_none`` attribute,
which supports more platforms and is more useful than LLVM's existing ``ghccc``
calling convention. This also removes the need to manually patch the calling
convention in LLVM IR, simplifying the JIT compilation process.
```

**Review by brandtbucher — COMMENTED** [review]
Looks good, just a few opportunities to tidy things up:

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-22-04-18-53.gh-issue-125498.cFjPIn.rst (hunk 9 above)
Comment (savannahostrowski) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-10-22-04-18-53.gh-issue-125498.cFjPIn.rst]: Ah yeah that's fair - it's kind of like ghccc but not quite (I was rifting on some of the language used in https://discourse.llvm.org/t/rfc-exposing-ghccc-calling-convention-as-preserve-none-to-clang/74233)

**Comment by savannahostrowski** [comment]
Oooo, thanks for the review @brandtbucher. Great catches for cleaning things up a bit. Much simpler.

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-22-04-18-53.gh-issue-125498.cFjPIn.rst (hunk 9 above)
Comment (brandtbucher) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-10-22-04-18-53.gh-issue-125498.cFjPIn.rst]: Yeah, they ended up changing it quite a bit in the end to use a more sane set of registers, haha.

**Review by brandtbucher — APPROVED** [review]
Awesome! @mdboom, I'm landing this, so we should probably also update the benchmark runners that need it. Let me know if you want help.

**Comment by mdboom** [comment]
> Awesome! @mdboom, I'm landing this, so we should probably also update the benchmark runners that need it. Let me know if you want help.

Yep -- I'll do that now.
